### PR TITLE
Zip code via URL

### PR DESCRIPTION
**Pull request by sebasblancogonz**
There it is a nice implementation to read zip code via the URL. The format would be something like this:
`localhost:5000/{zip_code}`

Here it is how it works:

1. It will check if there's any cache with the given zip code as a **key**. If it already exists, it will return the context that's been saved in that cache.

> `if cache.get(zip_code):
>         return render_template('index.html', **cache.get(zip_code))`
> 

   2. If there is not any cache with that key, it will execute the normal code, obtaining all the data from the API corresponding to the given zip code. Checks if the zip is `none` or not, if it's `none,` it will assign a default zip code, in this case, New York, Manhattan, Central Park.
After that it will follow the normal flux of the application, storing the new cache object with the zip code as the key.

It could be a nice approach, let me know if it works good for you!

### CODE REVIEW

**Comment by sebasblancogonz** [comment]
@JulianNorton you can check the modifications I've made on this PR. Caching is limited to 3 calls, so 3 unique locations would be cached. Should we rewrite the cached locations if new calls arrive?

**Comment by JulianNorton** [comment]
Thanks @sebasblancogonz. Right now I think if a 4th unique location is requested, it should overwrite the oldest cache. Let me review and get back to you!

File: app.py [hunk 1]
```diff
@@ -1,56 +1,88 @@
...
+            return render_template('index.html', **cache.get(CACHE_3_ZIP)['context'])
+        else:
+            timestamp = datetime.datetime.now()
+
+            lat, lng, default_value = 0, 0, False
+
+            if zip_code is None:
+                zip_code, default = 10019, True #New York, Manhattan, Central Park (default)
+
+            with open('./assets/data.csv', 'r') as f:
```
Comment (JulianNorton) [inline app.py]: where did you source the zipcodes and lat/long? Commenting with the source or how to verify would be useful to troubleshoot in the future. I also recommend renaming `data.csv` to something more semantic to reflect *what*  the data is. For example: `lat-long-zipcode.csv`

Comment (JulianNorton) [reply app.py]: I'm going to test this locally soon

**Comment by JulianNorton** [comment]
@sebasblancogonz I tried zipcode 01610 (Worcester, MA) and got a key error – can you look into it?

image

**Review by JulianNorton — CHANGES_REQUESTED** [review]
Encountered fatal errors in local review: zip 01610

**Comment by sebasblancogonz** [comment]
Going to check it out

**Comment by sebasblancogonz** [comment]
Hello Julian, the problem is that apparently for that given ZIP there is no available data on the following endpoint: **https://api.weather.gov/gridpoints/BOX/46,66/forecast**.
We have the following information for that ZIP:

> [zip_code: '01610', lat: '42.247049', lng: '-71.808366']

If we get the area_data for the given zip executing the call **https://api.weather.gov/points/42.92,-71.80**, we will get this forecast endpoint mentioned above, which will return the following data:

imagen

I will add some error handling and send some message to the front-end

File: app.py (hunk 1 above)
Comment (sebasblancogonz) [reply app.py]: > 
> 
> where did you source the zipcodes and lat/long? Commenting with the source or how to verify would be useful to troubleshoot in the future. I also recommend renaming `data.csv` to something more semantic to reflect _what_ the data is. For example: `lat-long-zipcode.csv`

Sure! Thanks for the advice, gonna change it right away!

Comment (sebasblancogonz) [reply app.py]: Sources are from here https://gist.github.com/abatko/ee7b24db82a6f50cfce02afafa1dfd1e

**Comment by sebasblancogonz** [comment]
With the new changes the app shouldn't crash when an invalid zip code is given. Now I'm working on the functionality to manage and rewrite the oldest cache! I'll let you know as soos as I got that done!

**Comment by JulianNorton** [comment]
Hey @sebasblancogonz,

I tried with a couple different zips and couldn't get it to work: `01610`, `10025`, and `91780` and get the not found message:
image


Is this something I'm not testing right? Or should I wait for your additional changes? I appreciate it and let me know if you need anything from me!

**Comment by sebasblancogonz** [comment]
I was wondering how could I do the cache rewriting stuff aand, what about queues? FIFO?

**Comment by JulianNorton** [comment]
I recommend to split it out into separate PRs and issues. Maybe it's best to just focus on the URL input first. Once that's confirmed working, we can work together on the cache?

**Comment by sebasblancogonz** [comment]
> 
> 
> I recommend to split it out into separate PRs and issues. Maybe it's best to just focus on the URL input first. Once that's confirmed working, we can work together on the cache?

Sure, that's perfect for me! I'll submit the last modification to avoid non-valid format zip codes

**Comment by sebasblancogonz** [comment]
Not it's finished! Test it and let me know if I have to change anything! 😊

File: app.py [hunk 2]
```diff
@@ -1,67 +1,129 @@
...
+
+    if zip_code != None and re.search(r'.*(\d{5}?)$', zip_code):
+        if cache.get(CACHE_1_ZIP) is not None and cache.get(CACHE_1_ZIP)['zip_code'] == zip_code:
+            return render_template('index.html', **cache.get(CACHE_1_ZIP)['context'])
+        elif cache.get(CACHE_2_ZIP) is not None and cache.get(CACHE_2_ZIP)['zip_code'] == zip_code:
+            return render_template('index.html', **cache.get(CACHE_2_ZIP)['context'])
+        elif cache.get(CACHE_3_ZIP)  is not None and cache.get(CACHE_3_ZIP)['zip_code'] == zip_code:
+            return render_template('index.html', **cache.get(CACHE_3_ZIP)['context'])
+        else:
+            lat, lng, default_value = 0, 0, False
```
Comment (JulianNorton) [inline app.py]: default_value isn't use anywhere? Setting this to a `false` is not needed because it's not referenced again

File: app.py [hunk 3]
```diff
@@ -1,67 +1,129 @@
...
+            
+            # zipcode-lat-lng.csv
+            # https://gist.github.com/abatko/ee7b24db82a6f50cfce02afafa1dfd1e
+            with open('./assets/zipcode-lat-lng.csv', 'r') as f:
```
Comment (JulianNorton) [inline app.py]: The zip code values aren't returning correct weather location data. For comparison, I'm trying to get `10025` to match.

https://forecast.weather.gov/MapClick.php?lat=40.7999209&lon=-73.96831020000002&site=all&smap=1&searchresult=New%20York%2C%20NY%2010025%2C%20USA

However, when I use zipcodes from the file, I get `West New York, NJ`

10018 should also be New York, but it's returning NJ. 

The forecasts are *very* close but there are differences.
image

This is likely an original bug in my default lat/long having it return a *close* New Jersey weather station instead of the specific New York one.

File: app.py [hunk 4]
```diff
@@ -1,67 +1,129 @@
...
+    return context
+
+@app.route('/')
+@app.route('/index')
+@app.route('/<zip_code>')
+def index(zip_code=None):
+    
+    timestamp = datetime.datetime.now()
+
+    if zip_code != None and re.search(r'.*(\d{5}?)$', zip_code):
```
Comment (JulianNorton) [inline app.py]: Please remove all new cache logic, let's tackle that in a separate pull request!

File: app.py [hunk 5]
```diff
@@ -1,67 +1,129 @@
...
+            def save_data(data):
+                if cache.get(CACHE_1_ZIP) is None:
+                    cache.set(CACHE_1_ZIP, data)
+                    return
+                elif cache.get(CACHE_2_ZIP) is None:
+                    cache.set(CACHE_2_ZIP, data)
+                    return
+                elif cache.get(CACHE_3_ZIP) is None:
+                    cache.set(CACHE_3_ZIP, data)
+                    return
```
Comment (JulianNorton) [inline app.py]: please separate cache logic into a separate PR/branch! That way we can focus on just the zip codes.

File: app.py [hunk 6]
```diff
@@ -1,67 +1,129 @@
...
+
+            save_data(data)
+
+            return render_template('index.html', **context)
+    else:
+        if zip_code is None:
+            #If no zip is provided, a default forecast is returned
+            #New York, Manhattan, Central Park (default)
+            return render_template('index.html', **getContext(40.76, -73.98, True, timestamp))
```
Comment (JulianNorton) [inline app.py]: This is my original bug – the lat/long is returning a NJ weather station instead of the `KNYC` weather station

File: app.py [hunk 7]
```diff
@@ -1,67 +1,129 @@
...
+        else:
+            context = {
+                    'error': True,
+                    'title': 'Something went wrong',
+                    'description': 'The following zipcode %s has an invalid format, it must be a 5 digit number.' %zip_code
```
Comment (JulianNorton) [inline app.py]: good error catching! tested with a random thing and it works.

File: constants.py [hunk 8]
```diff
@@ -0,0 +1,3 @@
+CACHE_1_ZIP = 'CACHE_1_ZIP'
+CACHE_2_ZIP = 'CACHE_2_ZIP'
+CACHE_3_ZIP = 'CACHE_3_ZIP'
```
Comment (JulianNorton) [inline constants.py]: same as above – remove new cache logic from this PR

**Review by JulianNorton — CHANGES_REQUESTED** [review]
The lookup time for the zipcodes might be too long. I'm getting about an additional 100ms on this branch than I am on master.

Screen Shot 2019-10-24 at 3 15 25 PM
Screen Shot 2019-10-24 at 3 16 05 PM


To be more clear, here are the exact criteria I'm hoping can be resolved in this PR:
1. Maintain sub 100ms load time locally for zipcode `10025` ❌
2. Have zipcode `10018` return same weather forecast as weather.gov for `10018`  ❌. I think this is due to incorrect lat/long coordinates.
3. Have zipcode `90210` return beverly hills CA (Working!) ✅ 
4. Have an invalid zipcode `sdgljkgsdfjklh` return an error ❌) 
5. Have an invalid zipcode `841298489124891` return an error ✅
6. Remove new cache logic ❌

Default zip 10025 is  returning NJ instead of NYC is an existing bug and I should make a new github issue for that.

**Comment by JulianNorton** [comment]
Stale PR
